### PR TITLE
Hash firmware partition and load proof bundle

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: binary_transparency/firmware/devices/usb_armory/bootloader/main.go [hunk 1]
```diff
@@ -54,24 +55,45 @@ func main() {
...
+		Offset: kernelOffset,
 	}
-
-	if err = conf.Read(partition, defaultConfigPath); err != nil {
+	if err = conf.Read(kernelPart, defaultConfigPath); err != nil {
 		haltAndCatchFire(fmt.Sprintf("invalid configuration: %v\n", err), 4)
 	}
 
+	usbarmory.LED("white", true)
+	h, err := hashPartition(89246720, kernelPart)
```
Comment (mhutchinson) [inline binary_transparency/firmware/devices/usb_armory/bootloader/main.go:76]: That number is a bit magic. Should it be passed in as an argument? (well, initialized at compile time like StartKernel)

**Review by mhutchinson — APPROVED** [review]

File: binary_transparency/firmware/devices/usb_armory/README.md [hunk 2]
```diff
@@ -108,8 +108,9 @@ To compile the bootloader itself, run the following command in the `bootloader`
 directory:
 
 ```bash
-# Note that START corresponds to the offset of the firmwware partition;
-make CROSS_COMPILE=arm-none-eabi- TARGET=usbarmory imx BOOT=uSD START=10753536
+# Note that START_KERNEL corresponds to the offset of the firmwware partition,
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply binary_transparency/firmware/devices/usb_armory/README.md:111]: looks like it :/

File: binary_transparency/firmware/devices/usb_armory/bootloader/Makefile [hunk 3]
```diff
@@ -14,7 +14,9 @@ BUILD = ${BUILD_USER}@${BUILD_HOST} on ${BUILD_DATE}
 REV = $(shell git rev-parse --short HEAD 2> /dev/null)
 
 SHELL = /bin/bash
-START ?= 5242880
+START_KERNEL ?= 5242880
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply binary_transparency/firmware/devices/usb_armory/bootloader/Makefile:17]: good call - done

File: binary_transparency/firmware/devices/usb_armory/bootloader/ft.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	var bundle api.ProofBundle
+	if err := json.Unmarshal(f, &bundle); err != nil {
+		return api.ProofBundle{}, fmt.Errorf("invalid bundle: %w", err)
+	}
+	return bundle, nil
+}
+
+// hashPartition calclates the SHA256 of the first numBytes of the partition.
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply binary_transparency/firmware/devices/usb_armory/bootloader/ft.go:31]: done

File: binary_transparency/firmware/devices/usb_armory/bootloader/main.go (hunk 1 above)
Comment (AlCutter) [reply binary_transparency/firmware/devices/usb_armory/bootloader/main.go]: Yeah, it _is_ a bit of a random number, but I've made it passable-in like the others for now.

File: binary_transparency/firmware/devices/usb_armory/bootloader/ft.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+func hashPartition(numBytes int, p *Partition) ([]byte, error) {
+	fmt.Println("Reading partition...")
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply binary_transparency/firmware/devices/usb_armory/bootloader/ft.go]: Done

File: binary_transparency/firmware/devices/usb_armory/bootloader/ft.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	bs := 1<<16
+	rc := make(chan []byte, 10)
+	hc := make(chan []byte)
+
+	go func() {
+		h := sha256.New()
+		for b := range(rc) {
+		if _, err := h.Write(b); err != nil {
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply binary_transparency/firmware/devices/usb_armory/bootloader/ft.go]: :/ fixed

**Pull request merged**